### PR TITLE
Write JSDoc for map, mapTo, pluck, filter, scan, do

### PR DESCRIPTION
**Pull request by staltz**

### CODE REVIEW

File: src/operator/do.ts [hunk 1]
```diff
@@ -4,14 +4,50 @@ import {Observable} from '../Observable';
...
+ * <span class="informal">Intercepts each emission on the source and runs a
+ * function, but returns an output which is identical to the source.</span>
+ *
+ * <img src="./img/do.png" width="100%">
+ *
+ * Returns a mirrored Observable of the source Observable, but modified so that
+ * the provided Observer is called for every item emitted by the source.
+ *
+ * This operator is useful for debugging your Observables for the correct values
+ * or performing other side effects.
```
Comment (blesh) [inline src/operator/do.ts]: I'd swap this around to say it's useful for side effects *first*, and _then_ mention it's useful for debugging.

Comment (blesh) [inline src/operator/do.ts]: Nevermind, it looks like that's covered above. That's what I get for skimming.

File: src/operator/do.ts [hunk 2]
```diff
@@ -4,14 +4,50 @@ import {Observable} from '../Observable';
...
- * This operator is useful for debugging your observables for the correct values or performing other side effects.
- * @param {Observer|function} [nextOrObserver] a normal observer callback or callback for onNext.
- * @param {function} [error] callback for errors in the source.
- * @param {function} [complete] callback for the completion of the source.
- * @reurns {Observable} a mirrored Observable with the specified Observer or callback attached for each item.
+ * Perform a side effect for every emission on the source Observable, but return
+ * an Observable that is identical to the source.
+ *
+ * <span class="informal">Intercepts each emission on the source and runs a
+ * function, but returns an output which is identical to the source.</span>
```
Comment (blesh) [inline src/operator/do.ts]: How about:

> Allows for "mid-stream" observation of an observable chain by registering an observer or handlers that to observe values, errors and completions, generally for the purpose of creating side effects. Passes all events from the source to the returned observable. Any errors that are thrown in aforementioned observer or handlers are safely sent down the error path of the returned observable.

Thoughts?

File: src/operator/do.ts [hunk 3]
```diff
@@ -4,14 +4,50 @@ import {Observable} from '../Observable';
...
+ * execution, it does not trigger an execution to happen like `subscribe` does.
+ *
+ * @example <caption>Map every every click to the clientX position of that click, while also logging the click event</caption>
+ * var clicks = Rx.Observable.fromEvent(document, 'click');
+ * var positions = clicks
+ *   .do(ev => console.log(ev))
+ *   .map(ev => ev.clientX);
+ * positions.subscribe(x => console.log(x));
+ *
+ * @example <caption>Equivalent to the example above, where `do` is replaced by `map`</caption>
```
Comment (blesh) [inline src/operator/do.ts]: What is the intent of showing `do` replaced by `map` here?

File: src/operator/do.ts (hunk 2 above)
Comment (staltz) [inline src/operator/do.ts]: Sure. I can add that somewhere, but I try to keep the `informal` description short (my rule of thumb so far is max 2 lines), so your suggestion is longer. I can put your sentence outside of this informal block, like under it. But in general the `informal` description is what I **verbally** explain to a work colleague, written down unaltered.

File: src/operator/do.ts (hunk 1 above)
Comment (staltz) [inline src/operator/do.ts]: ok :)

File: src/operator/do.ts (hunk 3 above)
Comment (staltz) [inline src/operator/do.ts]: To illustrate what `do` does, and that it's not anyhow magical. It just takes the input and returns it. But in between, may cause a side effect. If you want, I can remove it, no problem.

Comment (blesh) [inline src/operator/do.ts]: I think remove it. I wouldn't want to indirectly encourage people to use `map` for side effects.

File: src/operator/do.ts (hunk 2 above)
Comment (blesh) [inline src/operator/do.ts]: I think it's important to stress it's basically about creating a side effect and sending any thrown errors down the error path.

Comment (blesh) [inline src/operator/do.ts]: The "errors down the error path" part is the part most people miss about `do`.

**Comment by staltz** [comment]
@blesh Added "*Any errors that are thrown in the aforementioned Observer or handlers are safely sent down the error path of the output Observable.*" and removed the map example.

**Comment by blesh** [comment]
LGTM. Needs a rebase though.

**Comment by staltz** [comment]
Rebased. :white_check_mark:

**Comment by kwonoj** [comment]
Merged with f209294, thanks @staltz :)

**Pull request merged**